### PR TITLE
introduce contract stat for dydx derivatives.

### PR DESCRIPTION
**Pull request by teekenl**
- What is the purpose of this Pull Request?
  Introduce the contract stat for dydx perpetual derivatives. It will be listed on coingecko.com/en/derivatives and coingecko.com/en/exchanges/derivatives

- What is the related issue for this Pull Request (if this PR fixes issue, prepend with "Fixes" or "Closes")?
- [x] I have added Specs
- [x] (If implementing Market Ticker) I have verified that the `volume` refers to BASE
- [x] (If implementing Market Ticker) I have verified that the `base` and `target` is assigned correctly
- [x] I have implemented the `trade_page_url` method that links to the exchange page with the `base` and `target` passed in. If not available, enter the root domain of the exchange website.
- [x] I have verified at least **ONE** ticker volume matches volume shown on the trading page (use script below)

```
client = Cryptoexchange::Client.new
pairs = client.pairs 'exchange_name'
tickers = pairs.map do |p| client.ticker p end
sorted_tickers = tickers.sort_by do |t| t.volume end.reverse
```

### CODE REVIEW

**Comment by teekenl** [comment]
@ElvisLYC are you able to take a look at this PR. for spot trading, do we skip and return nothing for 
 `contract_stat` ?

File: lib/cryptoexchange/exchanges/dydx/services/contract_stat.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        def market_pairs_url
+          Cryptoexchange::Exchanges::Dydx::Services::Pairs::PAIRS_URL
+        end
+
+        def contract_info_url(market_pair, contract_type)
+          if contract_type == "perpetual"
+              "#{Cryptoexchange::Exchanges::Dydx::Market::API_URL}/perpetual-markets/#{market_pair.base}-#{market_pair.target}"
+          else
+            #TODO
+            nil
```
Comment (teekenl) [inline lib/cryptoexchange/exchanges/dydx/services/contract_stat.rb:39]: not sure what we do to handle the spot?

**Comment by ElvisLYC** [comment]
@teekenl my suggestion is:-

1) Before deployment, double check with the ticket assignor that all `tickers` listed here https://api.dydx.exchange/v1/stats/markets is referring to perpetual tickers (because it is stated as type `spot` in the api)

2)you don't need `dydx` and `dydx_perpetual` market, choose only one. referring to my comment above

File: lib/cryptoexchange/exchanges/dydx_perpetual/services/pairs.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        PAIRS_URL = "#{Cryptoexchange::Exchanges::DydxPerpetual::Market::API_URL}/stats/markets"
+
+        def fetch
+          output = super
+          adapt(output)
+        end
+
+        def adapt(output)
+          output['markets'].map do |pair, value|
+            next if value["type"] != "PERPETUAL"
```
Comment (ElvisLYC) [inline lib/cryptoexchange/exchanges/dydx_perpetual/services/pairs.rb:14]: verify that all tickers from `PAIRS_URL` is futures tickers then remove this line and fetch all tickers from `PAIRS_URL`

**Comment by ElvisLYC** [comment]
LGTM

File: lib/cryptoexchange/exchanges/dydx/services/market.rb [hunk 3]
```diff
@@ -19,20 +19,21 @@ def ticker_url
...
-          output['markets'].map do |ticker|
-            base, target = ticker[0].split('-')
+          output['markets'].map do |ticker, value|
+            next if value["type"] == "PERPETUAL"
+
+            base, target = ticker.split('-')
             market_pair = Cryptoexchange::Models::MarketPair.new(
               base: base,
               target: target,
               market: Dydx::Market::NAME
```
Comment (wenjieyek) [inline lib/cryptoexchange/exchanges/dydx/services/market.rb:29]: market_pair should add inst_id

**Pull request merged**